src/v/storage/segment.h

```diff
@@ -40,27 +40,70 @@ struct segment_closed_exception final : std::exception {
 class segment {
 public:
     using generation_id = named_type<uint64_t, struct segment_gen_tag>;
-    struct offset_tracker {
+    class offset_tracker {
+    public:
+        using committed_offset_t
+          = named_type<model::offset, struct committed_offset_tag>;
+        using dirty_offset_t
+          = named_type<model::offset, struct dirty_offset_tag>;
+        using stable_offset_t
+          = named_type<model::offset, struct stable_offset_tag>;
+
         offset_tracker(model::term_id t, model::offset base)
-          : term(t)
-          , base_offset(base)
-          , committed_offset(model::prev_offset(base))
-          , dirty_offset(model::prev_offset(base))
-          , stable_offset(model::prev_offset(base)) {}
-        model::term_id term;
-        model::offset base_offset;
+          : _term(t)
+          , _base_offset(base)
+          , _committed_offset(model::prev_offset(base))
+          , _stable_offset(model::prev_offset(base))
+          , _dirty_offset(model::prev_offset(base)) {}
+
+        template<typename... Ts>
+        void set_offsets(Ts&&... ts) {
+            (set_offset_impl(std::forward<Ts>(ts)), ...);
+            vassert(
+		    _committed_offset <= _stable_offset && _stable_offset <= _dirty_offset,
+              "Must maintain offset invariant: committed <= stable <= dirty");
+        }
+
+        template<typename T>
+        void set_offset(T&& t) {
+            set_offsets(std::forward<T>(t));
+        }
+
+        model::term_id get_term() const { return _term; }
+        model::offset get_base_offset() const { return _base_offset; }
+        model::offset get_committed_offset() const { return _committed_offset; }
+        model::offset get_stable_offset() const { return _stable_offset; }
+        model::offset get_dirty_offset() const { return _dirty_offset; }
+
+    private:
+        template<typename T>
+        void set_offset_impl(T&& tagged_offset) {
+            if constexpr (std::is_same_v<T, committed_offset_t>) {
+                _committed_offset = tagged_offset();
+            } else if constexpr (std::is_same_v<T, stable_offset_t>) {
+                _stable_offset = tagged_offset();
+            } else if constexpr (std::is_same_v<T, dirty_offset_t>) {
+                _dirty_offset = tagged_offset();
+            } else {
```

Comment:
would be great to have some unit tests that check this type magic is working as expected.